Rubberduck.Parsing/ComReflection/UserProjectRepository.cs

```diff
@@ -0,0 +1,155 @@
+﻿using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using NLog;
+using Rubberduck.Parsing.Common;
+using Rubberduck.Parsing.UIContext;
+using Rubberduck.Parsing.VBA.Extensions;
+using Rubberduck.VBEditor.ComManagement;
+using Rubberduck.VBEditor.ComManagement.TypeLibs;
+using Rubberduck.VBEditor.SafeComWrappers.Abstract;
+
+namespace Rubberduck.Parsing.ComReflection
+{
+    public class UserProjectRepository : IUserComProjectRepository
+    {
+
+        private readonly IUiDispatcher _uiDispatcher;
+        private readonly ITypeLibWrapperProvider _typeLibWrapperProvider;
+        private readonly IProjectsProvider _projectsProvider;
+
+        private readonly IDictionary<string, ComProject> _userComProjects = new Dictionary<string, ComProject>();
+
+        private readonly Logger _logger = LogManager.GetCurrentClassLogger();
+
+
+        public UserProjectRepository(ITypeLibWrapperProvider typeLibWrapperProvider, IUiDispatcher uiDispatcher,
+            IProjectsProvider projectsProvider)
+        {
+            _typeLibWrapperProvider = typeLibWrapperProvider;
+            _uiDispatcher = uiDispatcher;
+            _projectsProvider = projectsProvider;
+        }
+
+        //NOTE: Before this class can be used outside the parsing process thread-safety measures will have to be added.
+        public ComProject UserProject(string projectId)
+        {
+            return _userComProjects.TryGetValue(projectId, out var result) ? result : null;
+        }
+
+        public IReadOnlyDictionary<string, ComProject> UserProjects()
+        {
+            //We return a copy to avoid issues updating the collections contents. 
+            return _userComProjects.ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
+        }
+
+        public void RefreshUserComProjects(IReadOnlyCollection<string> projectIdsToReload)
+        {
+            var parsingStageTimer = ParsingStageTimer.StartNew();
+
+            RemoveNoLongerExistingProjects();
+            RemoveProjects(projectIdsToReload);
+            var loadTask = _uiDispatcher.StartTask(() =>
+            {
+                AddUnprotectedUserProjects(projectIdsToReload);
+                AddLockedProjects();
+            });
+            loadTask.Wait();
+
+            parsingStageTimer.Stop();
+            parsingStageTimer.Log("Loaded ComProjects for user projects in {0}ms.");
+        }
+
+        private void RemoveNoLongerExistingProjects()
+        {
+            var existingProjectIds = _projectsProvider.Projects().Select(tpl => tpl.ProjectId)
+                .Concat(_projectsProvider.LockedProjects().Select(tpl => tpl.ProjectId)).ToHashSet();
+            var noLongerExistingProjectIds = _userComProjects.Keys
+                .Where(projectId => !existingProjectIds.Contains(projectId))
+                .ToList();
+            RemoveProjects(noLongerExistingProjectIds);
+        }
+
+        //Use only on the UI thread.
```

Comment:
why is that the case?

Reply:
This method uses `TryLoadProject`, which can only safely be used on the UI thread.

Reply:
I think the comment should at least remind why - perhaps `//TypeLib API requires the code to be executed on UI thread only```